### ISSUE TITLE
Feature request: Save Document

### ISSUE THREAD

**Issue by hematinik** [penge/my-notes#238]
I think it would be nice if you could download the document directly from the my-notes panel. currently you have to go to the Google Drive in order to get an offline copy of your document.

the HTML format is perfectly fine as the downloaded file format however I think there might be some complications regarding the images (I think you'll have to convert image URLs to a local address) so maybe converting to PDF/DOCX before the download would saves us a trouble.

**Comment by penge** [comment]
Or there could be option to **export all notes** (Options page) into a Zip file.

**"Print"** can be used to save current note to PDF if needed.

**Comment by hematinik** [comment]
Yep **Print** will do (silly me!).. the **Export All Notes** is a good Idea and a handy shortcut for sure

**Comment by penge** [comment]
Just had a thought. We can have export, but why leave it there? How about import? Imagine you can drag and drop any TXT file to Sidebar, it auto-creates the note with the same name or first available name, with all the text already imported.

@hematinik What do you think?

**Comment by hematinik** [comment]
that would be great! 
do you think we can import cloud-based documents as well ? like you paste the URL of a saved document (in Google docs format) and we have it imported document inside my-notes?

P.S
sorry for my late response and huge thanks for the Image drag&drop feature! it works like charm

**Comment by penge** [comment]
PS: That's great to hear! Thank you.

No worries about a later response. No rush. So about that import from URL, that might work in some cases only:

- response is in text format so it can be parsed, like JSON
- resource can be downloaded from any origin ([CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin))

Will create a task though that could import dropped TXT file.

**Comment by hematinik** [comment]
About the JSON response Im not sure if Google provides such API. I was thinking about Integration with other Note servcies like Evernote and MS Onenote. say you are logged in in your Evernote account and you want to import some notes from there, what's the best alternative to a copy-paste?
as for the Google Docs you may save an offline copy and import it to my-notes, so maybe we can automate the process like save the document in memory (from provided URL) and then convert it to something that my-notes understands.
- - -
we had an open issue regarding the making notes read-only. IMO that feature has a higher priority however I think you closed the issue for some reason, is this related to a technical problem?

**Comment by penge** [comment]
I've checked Evernote and there are two options: Shareable link (a public link), Invite someone (by email who has Evernote). Seems like there is no export/import but I would say that is because their note (which is HTML btw) looks how it looks because of their editor. And if the same editor wouldn't be on the other side (where importing), the experience wouldn't be the same, so just to avoid that there is no export. And probably also because you'd like to keep users in Evernote, not provide options to pack it up and go elsewhere :D

Haven't checked MS Onenote but I think that might be similar. Copy-paste seems to be the only although not convenient way.

**Read-only note** - I haven't found that particularly useful but maybe I am wrong. I think I never needed to lock a note, not typing into it just did the same. Technical problem there is not, it is quite simple - adding new Lock/Unlock to the note context menu, not allowing any typing/pasting when it's locked, some visual indication when it's locked (maybe a lighter color or lock icon). The only more complicated thing would be to sync this status: locked/unlocked, across devices. But that is something that might not be necessary after all, as it's easy to lock/unlock on every device and sometimes having it locked on one but unlocked on other can be preferred.

TLDR: Read-only note is easy to have. Do we need it?

**Comment by hematinik** [comment]
Yes the Evernote provides a lot of formatting options and in my experience this is the root of devil when dealing with large notes. Evernote keeps the exact formatting (and font face) when you copy-paste some web content and soon you'll end up with a sloppy note, imagine what happens when you want to add/edit something inside the note and it will inherit the font face and format from the previous line (I get hysterical just by thinking about it LOL)

the best user experience comes when you keep it simple and consistent (that's why I love my-notes), as a mindset I tend to avoid over-development (less fancy new features - more enhancements)

- - -

regarding the read-only option, personally it is a life saver for me, and generally I think many people suffer accidental inputs/unwanted pasting/etc.. specially when you are in rush and you care about the correctness of your note.

**Comment by penge** [comment]
Yeah, copy-pasting HTML from sites attaches styles inline and it gets quickly broken when editing. I try keep as much semantic styles separate from note as possible, so it could be customized in many details, like highlighter pen which can be customized for every theme, it doesn't have to be highlight as usual, it could be very thick underline too. So I try to put all those styles out. And also not use any generated CSS classes, as that's pain to locate and customize.

I am glad you like my notes!

Cool, I'll reopen read-only / lock, why not. I think it could be added as soon as next version.

**Comment by hematinik** [comment]
Im glad to hear that..! thanks a lot!
